### DISCUSSION TITLE
UI and multiple windows support/alternatives?

### DISCUSSION THREAD

**Discussion by ArcticXWolf** [hajimehoshi/ebiten#1831] [Q&A]
Hey,

first of all thanks for this great project, it really helps a lot and I hope we can get more and more people to use golang for game development!

I am currently developing a [Gameboy Color emulator](https://github.com/ArcticXWolf/AXWGameboy) and use ebiten to handle display, input and audio. My target platforms are:

* Windows
* Linux
* Android

Everything works really well, the emulation is 90% done (most games work) and it is also fast enough on all target platforms. However I now struggle to implement several "User Experience"-Features. Especially two aspects I do not know how to solve with ebiten:

1. **How do I implement UI elements?** I need to allow the user to select the game ROM (so file input), handle settings like keybindings (at least buttons and checkboxes), maybe add cheatcodes (so text input), etc. However no library supports all of that on my target platforms (especially text input and file selection on mobile seems to be a problem). I've seen [ebitenUI](https://github.com/blizzy78/ebitenui), but it is not for mobile. I know this is not part of the scope of ebiten and I probably need to implement that stuff myself, but I wanted to ask if someone here has had a similar issue and found a solution already.

And especially **help to create text input with gomobile** would be awesome, because AFAIK there is no way to trigger the android system keyboard from golang/gomobile. (File input I have been able to solve via a custom androidManifest and creating simple buttons for all files in question).

2. How do I create multiple windows? A lot of debugging for emulators consists of drawing internal state apart from the current game display like tilemaps, spritelists, etc. For that it would be awesome to be able to create a separate window to draw that stuff onto. This does not have to be cross plattform for me, just desktop is enough. I tried to work around the single window restriction via "subimages", but they are quite the hassle to organize in code IMHO. Is there a way in ebiten to facilitate this?

Thanks in advance on any tips or links and thanks again for such a cool project :)

**Comment by ArcticXWolf** [comment]
Oh and concerning the first point:
I played with the idea of using webassembly as the main target platform as it is cross platform and via HTML you get all those UI elements for free. However:
1. during my initial tests the performance was worse than the native implementation
2. the interaction between JS and golang was a bit difficult for me (Especially if served via iFrame: how do I interact with golang INSIDE the iframe?)
3. I have to host the emulator somewhere (or the user needs to do it themselves). A native executable is much nicer...

**Comment by hajimehoshi** [comment]
Hi, thank you for using Ebiten!

1. You'd need to implement text input in Java by yourself unfortunately. There is a plan to implement a native text input for each environment, but there is still a long way before it is implemented.
2. There is no way to handle multiple windows. This requires a lot of design and implement change in Ebiten. I recommend to create multiple Ebiten processes so far, but this might be complicated.

**Reply by ArcticXWolf** [reply]
Thanks for the fast reply!

Yeah, I thought as much, but I will keep this threat open in case that someone finds something new.

Probably Webassembly is the way to go, then I can package it inside a PWA or such. On that note: do you know if it is possible for the HTML and JS to interact with the webassembly running in the iFrame? Or do I need to implement everything also inside the iframe (aka spawn the corresponding code via syscall/js)?

**Reply by hajimehoshi** [reply]
It's possible to interact with an iframe by [`window.postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage).
